plan/logical_plan_builder.go

```diff
@@ -409,6 +411,74 @@ func (b *planBuilder) buildSelection(p LogicalPlan, where ast.ExprNode, AggMappe
 	return selection
 }
 
+// buildProjectionFieldNameForColumns builds the field name and the table name when field expression is a column reference.
+func (b *planBuilder) buildProjectionFieldNameForColumns(field *ast.SelectField, c *expression.Column) (colName model.CIStr, tblName model.CIStr) {
+	if astCol, ok := getInnerFromParentheses(field.Expr).(*ast.ColumnNameExpr); ok {
+		colName = astCol.Name.Name
+		tblName = astCol.Name.Table
+	} else {
+		colName = c.ColName
+		tblName = c.TblName
+	}
+	return
+}
+
+// buildProjectionFieldNameForExpressions builds the field name when field expression is a normal expression.
+func (b *planBuilder) buildProjectionFieldNameForExpressions(field *ast.SelectField) (colName model.CIStr) {
+	if agg, ok := field.Expr.(*ast.AggregateFuncExpr); ok && agg.F == ast.AggFuncFirstRow {
+		// When the query is select t.a from t group by a; The Column Name should be a but not t.a;
+		colName = agg.Args[0].(*ast.ColumnNameExpr).Name.Name
+	} else {
+		// Other kind of (normal) expressions.
+		innerExpr := getInnerFromParentheses(field.Expr)
+
+		// Expression is a value literal.
+		if valueExpr, ok := innerExpr.(*ast.ValueExpr); ok {
+			switch valueExpr.Kind() {
+			case types.KindString:
+				// See #3686, #3994:
+				// For string literals, string content is used as column name. Non-graph initial characters are trimmed.
+				colName = model.NewCIStr(strings.TrimLeftFunc(valueExpr.GetString(), func(r rune) bool {
+					return !unicode.IsOneOf(mysql.RangeGraph, r)
+				}))
+			case types.KindNull:
+				// See #4053, #3685
+				colName = model.NewCIStr("NULL")
+			default:
+				if innerExpr.Text() != "" {
+					colName = model.NewCIStr(innerExpr.Text())
+				} else {
+					colName = model.NewCIStr(field.Text())
+				}
+			}
+		} else {
+			// Remove special comment code for field part, see issue #3739 for detail.
+			colName = model.NewCIStr(parser.SpecFieldPattern.ReplaceAllStringFunc(field.Text(), parser.TrimComment))
+		}
+	}
+	return
+}
+
+// buildProjectionField builds the field object according to SelectField in projection.
+func (b *planBuilder) buildProjectionField(field *ast.SelectField, fieldAsExpr expression.Expression) *expression.Column {
```

Comment:
1. add the `Projection`'s id and the field position as the first or last parameters of this function
2. s/fieldAsExpr/expr/

```go
func (b *planBuilder) buildProjectionField(id string, position int, field *ast.SelectField, expr expression.Expression) *expression.Column {
```